### PR TITLE
User permission

### PR DESCRIPTION
**Pull request by andrewseamanco**
Add user permissions
-Add permission enum
-Implement new enum in profile.jsp to only show requests.html to admin
-Disallow users with user permission to access requests.html

### CODE REVIEW

File: bait-files/src/main/java/com/google/LoginFilter.java [hunk 1]
```diff
@@ -71,20 +72,41 @@ else if (!userIsRegistered(userService.getCurrentUser().getUserId())
         RequestDispatcher requestDispatcher = request.getRequestDispatcher("WEB-INF/profile.jsp");
         requestDispatcher.forward(request, response);
         return;
+      } else if (request.getRequestURI().endsWith("requests.html")) {
+          System.out.println("Here");
```
Comment (kzweller) [inline bait-files/src/main/java/com/google/LoginFilter.java:76]: Remove.

File: bait-files/src/main/java/com/google/LoginFilter.java [hunk 2]
```diff
@@ -71,20 +72,41 @@ else if (!userIsRegistered(userService.getCurrentUser().getUserId())
         RequestDispatcher requestDispatcher = request.getRequestDispatcher("WEB-INF/profile.jsp");
         requestDispatcher.forward(request, response);
         return;
+      } else if (request.getRequestURI().endsWith("requests.html")) {
```
Comment (kzweller) [inline bait-files/src/main/java/com/google/LoginFilter.java:75]: Maybe a "Case" comment, to be consistent?

File: bait-files/src/main/java/com/google/LoginFilter.java [hunk 3]
```diff
@@ -71,20 +72,41 @@ else if (!userIsRegistered(userService.getCurrentUser().getUserId())
...
+          System.out.println("Here");
+          if (getCurrentUserPermission() == Permission.ADMIN) {
+              chain.doFilter(req, res);
+          } else {
+            RequestDispatcher requestDispatcher = request.getRequestDispatcher("WEB-INF/profile.jsp");
+            requestDispatcher.forward(request, response);  
+          }
       } else {
+        System.out.println(request.getRequestURI());
```
Comment (kzweller) [inline bait-files/src/main/java/com/google/LoginFilter.java:84]: Remove.

File: bait-files/src/main/java/com/google/LoginFilter.java [hunk 4]
```diff
@@ -71,20 +72,41 @@ else if (!userIsRegistered(userService.getCurrentUser().getUserId())
...
       }
     }
   }
-  private boolean userIsRegistered(String id) {
+
+  private boolean isRegistered(String id) {
     List<User> allUsers = ObjectifyService.ofy().load().type(User.class).list();
 
-    List<User> listOfUsersWithId =
+    List<User> isRegistered =
```
Comment (kzweller) [inline bait-files/src/main/java/com/google/LoginFilter.java:95]: This shouldn't change?

File: bait-files/src/main/java/com/google/LoginFilter.java [hunk 5]
```diff
@@ -71,20 +72,41 @@ else if (!userIsRegistered(userService.getCurrentUser().getUserId())
...
         allUsers.stream().filter(user -> user.getUserId().equals(id)).collect(toList());
 
-    return listOfUsersWithId.size() >= 1;
+    return isRegistered.size() >= 1;
+  }
+
+  private Permission getCurrentUserPermission() {
```
Comment (kzweller) [inline bait-files/src/main/java/com/google/LoginFilter.java:101]: Comment for what this is doing please.

File: bait-files/src/main/java/com/google/sps/data/Constants.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package com.google.sps.data;
```
Comment (kzweller) [inline bait-files/src/main/java/com/google/sps/data/Constants.java:1]: I thought one of these files existed already? Does it not? If not, can you look through the constants being used elsewhere and consolidate them using this file?

File: bait-files/src/main/java/com/google/sps/servlets/UserHistoryServlet.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Servlet loads all pending requests from datastore.
+ * Loads a requests and filters requests by status. Used to show all pending requests for review
+ * queue.
+ */
+
+@WebServlet("/userHistory")
```
Comment (kzweller) [inline bait-files/src/main/java/com/google/sps/servlets/UserHistoryServlet.java:25]: This change seems unrelated to "user permission"...?

File: bait-files/src/main/java/com/google/LoginFilter.java (hunk 1 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/LoginFilter.java]: Removed

File: bait-files/src/main/java/com/google/LoginFilter.java (hunk 4 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/LoginFilter.java]: Changed back

File: bait-files/src/main/java/com/google/sps/data/Constants.java (hunk 6 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/sps/data/Constants.java]: Talked to Iman and this is the only Java constants for right now

File: bait-files/src/main/java/com/google/sps/servlets/UserHistoryServlet.java (hunk 7 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/sps/servlets/UserHistoryServlet.java]: Merging removed this file

File: bait-files/src/main/java/com/google/LoginFilter.java (hunk 3 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/LoginFilter.java]: Removed

File: bait-files/src/main/java/com/google/LoginFilter.java (hunk 5 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/LoginFilter.java:105]: Added comment

Comment (kzweller) [reply bait-files/src/main/java/com/google/LoginFilter.java:105]: I meant.. more specifically. Why is it setting userPermission, just to immediately override it? Suggest something like...

```
return allUsers.stream()
   .filter(user -> 
      user.getUserId().equals(UserServiceFactory.getUserService().getCurrentUser().getUserId()))
  .collect(onlyElement())
  .getPermission();
```

File: bait-files/src/main/webapp/WEB-INF/profile.jsp [hunk 8]
```diff
@@ -10,13 +10,29 @@
...
+
+    <%
+        List<User> allUsers = ObjectifyService.ofy().load().type(User.class).list();
+        Permission userPermission = Permission.USER;
+        for (User user : allUsers) {
+            if (user.getUserId().equals(UserServiceFactory.getUserService().getCurrentUser().getUserId())) {
+                userPermission = user.getPermission();
+            }
+        }
+    %>
```
Comment (kzweller) [inline bait-files/src/main/webapp/WEB-INF/profile.jsp:26]: Same comment as above.

Comment (andrewseamanco) [reply bait-files/src/main/webapp/WEB-INF/profile.jsp]: Removed

File: bait-files/src/main/java/com/google/LoginFilter.java (hunk 5 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/LoginFilter.java:106]: Got it!

File: bait-files/src/main/java/com/google/sps/servlets/UserHistoryServlet.java [hunk 9]
```diff
@@ -24,32 +26,47 @@
 
 @WebServlet("/userHistory")
 public class UserHistoryServlet extends HttpServlet {
+  class SortByMostRecentDate implements Comparator<Review> {
```
Comment (kzweller) [inline bait-files/src/main/java/com/google/sps/servlets/UserHistoryServlet.java]: I don't think you need this.. see suggestion below

File: bait-files/src/main/java/com/google/sps/servlets/UserHistoryServlet.java [hunk 10]
```diff
@@ -24,32 +26,47 @@
...
     }
   }
+
+  public List<Review> getReviewsByType(Status statusType) {
+    List<Review> allRequests = ObjectifyService.ofy().load().type(Review.class).list();
+    List<Review> reviews = allRequests.stream()
+                               .filter(rev -> rev.userId.equals(userId))
+                               .filter(rev -> rev.status == statusType)
+                               .collect(toList());
+    Collections.sort(reviews, new SortByMostRecentDate());
```
Comment (kzweller) [inline bait-files/src/main/java/com/google/sps/servlets/UserHistoryServlet.java]:
```suggestion
    List<Review> reviews = allRequests.stream()
                               .filter(rev -> rev.userId.equals(userId))
                               .filter(rev -> rev.status == statusType)
                               .sorted(Comparator.comparing(rev -> rev.submissionDate)
                               .collect(toList());
```

And if that's backwards then throw on ".reversed()" to the Comparator.

File: bait-files/src/main/java/com/google/sps/servlets/UserHistoryServlet.java (hunk 9 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/sps/servlets/UserHistoryServlet.java]: Changed to default comparator

File: bait-files/src/main/java/com/google/sps/servlets/UserHistoryServlet.java (hunk 10 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/sps/servlets/UserHistoryServlet.java]: Changed to default comparator
